app/ts/Subproblem/SubproblemButtons/AddSubproblemButton/AddSubproblemUtil.ts

```diff
@@ -0,0 +1,321 @@
+  items: IRelativePerformance[],
+  dataSourceId: string
+): IRelativePerformance {
+  return _.find(items, (item: IRelativePerformance) => {
+    return item.dataSourceId === dataSourceId;
+  });
+}
+
+function hasPerformance(performance: Effect | Distribution) {
+  return (
+    performance && performance.type !== 'empty' && performance.type !== 'text'
+  );
+}
+
+export function initInclusions<T extends {id: string}>(
+  items: Record<string, T>,
+  exclusions?: string[]
+): Record<string, boolean> {
+  return _.mapValues(items, (item: T) => {
+    return !exclusions || !_.includes(exclusions, item.id);
+  });
+}
+
+export function getScaleBlockingWarnings(
+  criterionInclusions: Record<string, boolean>,
+  dataSourceInclusions: Record<string, boolean>,
+  alternativeInclusions: Record<string, boolean>,
+  workspace: IWorkspace
+): string[] {
+  let warnings: string[] = [];
+  if (
+    areValuesMissingInEffectsTable(
+      workspace,
+      dataSourceInclusions,
+      alternativeInclusions
+    )
+  ) {
+    warnings.push('Effects table contains missing values');
+  }
+  if (
+    areTooManyDataSourcesSelected(
+      _.keyBy(workspace.criteria, 'id'),
+      criterionInclusions,
+      dataSourceInclusions
+    )
+  ) {
+    warnings.push('Effects table contains multiple data sources per criterion');
+  }
+  return warnings;
+}
+
+function areValuesMissingInEffectsTable(
+  workspace: IWorkspace,
+  dataSourceInclusions: Record<string, boolean>,
+  alternativeInclusions: Record<string, boolean>
+): boolean {
+  const includedDataSourceIds = _.keys(_.pickBy(dataSourceInclusions));
+  const includedAlternativeIds = _.keys(_.pickBy(alternativeInclusions));
+
+  return _.some(includedDataSourceIds, (dataSourceId) => {
+    return _.some(includedAlternativeIds, (alternativeId) => {
+      const [effect, distribution, relativePerformance] = findPerformances(
+        workspace,
+        dataSourceId,
+        alternativeId
+      );
+
+      return (
+        !hasPerformance(effect) &&
+        !hasPerformance(distribution) &&
+        !relativePerformance
+      );
+    });
+  });
+}
+
+function areTooManyDataSourcesSelected(
+  criteria: Record<string, ICriterion>,
+  criterionInclusions: Record<string, boolean>,
+  dataSourceInclusions: Record<string, boolean>
+): boolean {
+  const includedCriterionIds = _.keys(_.pickBy(criterionInclusions));
+  return _.some(includedCriterionIds, (criterionId) => {
+    const criterion = criteria[criterionId];
+    const numberOfSelectedDataSources = getNumberOfSelectedDataSources(
+      criterion.dataSources,
+      dataSourceInclusions
+    );
+    return numberOfSelectedDataSources > 1;
+  });
+}
+
+function getNumberOfSelectedDataSources(
+  dataSources: IDataSource[],
+  dataSourceInclusions: Record<string, boolean>
+) {
+  return _.countBy(dataSources, (dataSource) => {
+    return dataSourceInclusions[dataSource.id];
+  }).true;
+}
+
+export function isAlternativeDisabled(
```

Comment:
`isAlternativeDeselectionDisabled` (see below for datasource equivalent function name)